### PR TITLE
Better support for resources in different namespaces

### PR DESCRIPTION
**Pull request by treydock**
# Pull Request Checklist

## Description

Resources that can belong to a namespace now support composite names in form of "<name> in <namespace>". This change is backwards compatible. I did have to change how some properties are validated as doing `validate do` for an entire type was not working with composite names. The official Puppet docs for custom types only document the use of `pre_run_check` which essentially does the same thing, performs validation after the catalog is compiled but before the catalog is applied. This is how properties are validated against other properties.

This also resolves a bug where it would have been impossible to delete a resource if it was not in the `default` namespace.

## Related Issue

Partial fix for #1125 

## Motivation and Context

In Sensu Go resources that belong to namespaces can have the same name so long as they are in different namespaces. This change ensures the Puppet types support this same behavior by matching resources by name + namespace and not just name.

### CODE REVIEW

File: spec/unit/sensu_role_spec.rb [hunk 1]
```diff
@@ -28,6 +28,21 @@
     }.to raise_error(Puppet::Error, 'Title or name must be provided')
   end
 
+  it 'should handle composite title' do
```
Comment (ghoneycutt) [inline spec/unit/sensu_role_spec.rb]: what happens if a composite title and a namespace are specified? what happens if they are specified incorrectly, ie `foo in qa` is the title and the namespace is `default`.

what happens if the title is named funny? `in in qa` or `weird       spacing                   in qa` or they use hard tabs?

File: README.md [hunk 2]
```diff
@@ -482,6 +483,33 @@ class { '::sensu::backend':
...
+All resources that support having a `namespace` also support a composite name to define the namespace.
+
+For example, the `sensu_check` with name `check-cpu in qa` would be named `check-cpu` and put into the `qa` namespace.
+
+Using composite names is necessary if you wish to have multiple resources with the same name but in different namespaces.
+
+For example to define the same check in two namespaces using the same check name:
+
+```puppet
+sensu_check { 'check-cpu in default':
```
Comment (ghoneycutt) [inline README.md]: if you don't specify a composite name would the namespace be specified in the type/provider as `default` or would it be unspecified?

File: spec/unit/sensu_role_spec.rb (hunk 1 above)
Comment (treydock) [reply spec/unit/sensu_role_spec.rb]: For the first issue, `namespace` property takes priority but I'll add a test for that. So name `foo in qa` with `namespace` of `default` will keep namespace `default`.

If the pattern doesn't match it just puts the entire name in to `name` property and that becomes default for `resource_name` and `resource_name` doesn't allow spaces so those 2 examples would raise an error. We already test for this indirectly but I'll see if I can add some reasonable tests for this with compound aspect.

File: README.md (hunk 2 above)
Comment (treydock) [reply README.md]: The default `namespace` is always `default` so if the name is just `check-cpu` the namespace is `default` if not specified.  I put the default in the name as thought it'd be confusing to mix non-composite and composite in the same example.

File: spec/unit/sensu_role_spec.rb (hunk 1 above)
Comment (ghoneycutt) [reply spec/unit/sensu_role_spec.rb]: Thanks! Since this introduces a regex on the title, we want to ensure we are testing the regex as well as the edge cases of people mixing composite naming with namespaces.

Mixing could happen inadvertently with Hiera pretty easily.

Comment (treydock) [reply spec/unit/sensu_role_spec.rb]: Added some extra unit tests for composite names.

File: README.md [hunk 3]
```diff
@@ -482,6 +483,33 @@ class { '::sensu::backend':
 }
 ```
 
+### Composite Names for Namespaces
+
+All resources that support having a `namespace` also support a composite name to define the namespace.
+
+For example, the `sensu_check` with name `check-cpu in qa` would be named `check-cpu` and put into the `qa` namespace.
+
+Using composite names is necessary if you wish to have multiple resources with the same name but in different namespaces.
```
Comment (ghoneycutt) [inline README.md]: Add something about if you use composite namespaces and specify `namespace` that `namespace` will take precedence. 

Perhaps also some guidance that if you use namespaces then you should follow this composite pattern. Ideally you wouldn't use namespaces for different environments, you would use hiera data. Is there another use case for the namespace? Don't want to confuse people.

Comment (treydock) [reply README.md]: I added a note about `namespace` property taking precedence.

I think namespaces can probably be different than puppet environments and unrelated to Hiera. Because roles define namespaces it's easy to imagine giving access controls to different namespaces that would all be defined in one Hiera environment so that the production sensu-backend could manage multiple namespaces for different roles or entities. As for composite vs non-composite, it's eluded to in README that the main benefit to composite is that you can have something like a check have the same name is Sensu for different environments. Besides having the same name in Sensu, there's not real benefit to using composite names if sites decide they want namespaces to be in the name they could just name checks `check-cpu-qa` or something.

Comment (ghoneycutt) [reply README.md]: so for the example should we change `qa` to something like `tenant1` or `namespace_foo`? I don't want to encourage folks to use namespace's for their deployment environments.

Comment (treydock) [reply README.md]: Ya I can change that to be something unique to a team or geographical location or something not specific to a sort of development workflow.

Comment (treydock) [reply README.md]: Updated examples in README.

**Comment by ghoneycutt** [comment]
Looks great. Ready for merge once Travis is happy.

**Comment by treydock** [comment]
Travis appears to be good now.
